client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java

```diff
@@ -38,6 +38,16 @@
 import org.apache.celeborn.common.util.ThreadUtils;
 import org.apache.celeborn.reflect.DynMethods;
 
+/*
+ In order to support Spark Stage resubmit with ShuffleReader FetchFails, Celeborn shuffleId has to be distinguished
+ from Spark shuffleId. Spark shuffleId is assigned at ShuffleDependency construction time, and all Attempts of
+ a Spark Stage have the same ShuffleId. When Celeborn ShuffleReader fails to fetch shuffle data from worker and
+ throws {@link FetchFailedException}, Spark DAGScheduler resubmits the failed ResultStage and corresponding ShuffleMapStage
+ , but Celeborn can't differentiate shuffle data from previous failed/resubmitted ShuffleMapStage with the same shuffleId.
+ Current solution takes Stage retry in account, and set
+         celeborn shuffleId = spark shuffleId * maxStageAttemptsNum + stageAttemptsId
```

Comment:
I probably mentioned this in the discussion in dev list - we cannot assume `spark shuffleId * maxStageAttemptsNum + stageAttemptsId` does not lead to conflicts.

With shuffle id reuse across jobs, this will result in conflicts.

Reply:
Mridul, as I explained in maillist, spark shuffle id is a static concept bounded to the shuffle dependency, aka, the shuffle data. I prefer not to call it "shuffle id reuse", since it is not the same id pointing to different data, it is just THE shuffle data's id. I don't think there will be any conflict.
If spark shuffle data 0 needs to be accessed, no matter it will be submitted to DAGScheduler soon, or it had been computed in previous job, spark shuffle id 0 should be used. In second case, spark read the shuffle data directly if the data doesn't need to be re-computed. you can search 'spark_shuffleId 0' in logs pasted in my mail to see that, shuffle data 0 is written once and read twice with spark shuffle id 0/celeborn shuffle id 0

Reply:
> I probably mentioned this in the discussion in dev list - we cannot assume `spark shuffleId * maxStageAttemptsNum + stageAttemptsId` does not lead to conflicts.
> 
> With shuffle id reuse across jobs, this will result in conflicts.

Mridul, as I explained in maillist, spark shuffle id is a static concept bounded to the shuffle dependency, aka, the shuffle data. I prefer not to call it "shuffle id reuse", since it is not the same id pointing to different data, it is just THE shuffle data's id. I don't think there will be any conflict.
If spark shuffle data 0 needs to be accessed, no matter it will be submitted to DAGScheduler soon, or it had been computed in previous job, spark shuffle id 0 should be used. In second case, spark read the shuffle data directly if the data doesn't need to be re-computed. you can search 'spark_shuffleId 0' in logs pasted in my mail to see that, shuffle data 0 is written once and read twice with spark shuffle id 0/celeborn shuffle id 0

Reply:
Spark uses shuffle id, stage reexecution, job reexecution (particularly for sql) in a specific way, and that does not change due to Celeborn being used.

In this specific context, the assumption being made is, for a given Spark shuffle id, there are upto maxStageAttemptNums stage attempts possible - if there is more, the generated id will conflict with some other Spark shuffle id's Celeborn shuffle id - given this, the question is, can this happen - and as I have illustrated before - it can, and does actually.

Please let me know if I am missing something etching or can help clarify

Reply:
To add, `spark.stage.maxConsecutiveAttempts` is for a specific stage, not shuffle id - so total number of times a particular shuffle id is compted by stage attempts has no bounds (across jobs)

Reply:
Thanks for the clarification, you are correct.
I feel the shuffle id mapping by (maxStageAttemptsNum, stageAttemptsId) is a bad idea now
The bounded shuffle id range enforced by maxStageAttemptsNum conflicts with the unbounded spark shuffle id range generated by spark stage re-run, which may happens when shuffle data gets recomputed with several round of shuffle fetch failure

Let me review the solution of singleton shuffle id generation in LifecycleMananger, and keep track of the correct shuffleId with
`private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[Int, Boolean]]()`

Reply:
Please do let me know if I can give help or give clarity ... I don't understand Celeborn well yet (particularly given it supports multiple engines), but I can help explain how Spark might behave for your proposals !

Reply:
basic idea of shuffle id generation in LifecycleManager:
1. shuffle writer sends app shuffle id to LifecycleManager, which may create new shuffle id for writing
- If the app shuffle id is not recorded in shuffleIdMapping, then generates a new shuffle id with AtomicInteger, puts appShuffleId -> (shuffleId -> true) in shuffleIdMapping, returns shuffleId to shuffle writer 
- If the app shuffle id is mapped to (shuffleId -> true), returns shuffleId to shuffle writer 
- If the app shuffle id is mapped to (shuffleId -> false), generates a new (shuffleId' -> true) to put into shuffleIdMapping, returns shuffleId' to shuffle writer // fetch failure happened

2. shuffle reader sends app shuffle id to LifecycleManager, gets latest shuffle id associated with app shuffle id 

3. during FetchFailure, shuffle reader sends ReportFetchFailure with (app shuffleId, shuffle id) to LifecycleManager, updates shuffleIdMapping to appShuffleId -> (shuffleId -> false)

@mridulm 
Do you see any issue here?

Reply:
To clarify, the incoming request to LifecycleManager will include app-shuffle-id, stage-id, stage-attempt-id ?
If yes, agree, this should address all the concerns I was having.

One small nit, if we can generate this at driver and ship the value to executor (so that we can avoid the invocation from executor to driver to fetch this id), that would be great - I have not thought through how we can do this, but adding as a comment in case there is an existing solution we can leverage !

Reply:
It is definitely a good idea to use app-shuffle-id + stage-id + stage-attempt-id as an identifier for one execution of shuffle computing, which maps to a celeborn shuffle id. let's call it appShuffleIdentifier
`val appShuffleIdentifier = s"$appShuffleId-$stageId-$stageAttemptId"`

the solutions is updated as:
define 
```
  // app shuffle id -> LinkedHashMap of (app shuffle identifier, (shuffle id, fetch status))
  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[String, (Int, Boolean)]]()
```
1. shuffle writer sends (appShuffleId, appShuffleIdentifier, isWriter=true) to LifecycleManager

- If appShuffleId or appShuffleIdentifier is not recorded, then generates a new shuffle id with AtomicInteger, saves mapping of appShuffleId -> appShuffleIdentifier -> (shuffleId, true) and return to writer
- if appShuffleIdentifier is found, return shuffleId to writer

2. shuffle reader sends (appShuffleId, null, isWriter=false) to LifecycleManager

- find the latest finished(all shuffle write task is done in celeborn) shuffle id associated with appShuffleId, return to reader

3. during FetchFailure, shuffle reader sends ReportFetchFailure with (app shuffleId, shuffle id) to LifecycleManager, cleans status on MapOutputTracker (exactly once) and updates fetch status of the shuffle id to false, and does nothing if fetch status is already false

Regarding 2, reader needs to get the latest finished shuffleId to prevent reader from a previous job to incorrectly get the running resubmitted ShuffleMapStage's shuffleId (that is unfinished) and reports fetch failure mistakenly